internal/controller/controllers/agent_controller.go

```diff
@@ -238,6 +214,61 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	return r.updateStatus(ctx, log, agent, origAgent, &h.Host, h.ClusterID, nil, false)
 }
 
+func (r *AgentReconciler) handleAgentFinalizer(ctx context.Context, log logrus.FieldLogger, agent *aiv1beta1.Agent) (*ctrl.Result, error) {
+	if agent.ObjectMeta.DeletionTimestamp.IsZero() { // agent not being deleted
+		// Register a finalizer if it is absent.
+		if !funk.ContainsString(agent.GetFinalizers(), AgentFinalizerName) {
+			controllerutil.AddFinalizer(agent, AgentFinalizerName)
+			if err := r.Update(ctx, agent); err != nil {
+				log.WithError(err).Errorf("failed to add finalizer %s to resource %s %s", AgentFinalizerName, agent.Name, agent.Namespace)
+			}
+			// After update there should not be any more changes on the object
+			// Update will return a new object so the creation of maps like annotations or labels is not valid anymore
+			return &ctrl.Result{Requeue: true}, nil
+		}
+	} else { // agent is being deleted
+		if funk.ContainsString(agent.GetFinalizers(), AgentFinalizerName) {
+			if _, has_annotation := agent.GetAnnotations()[BMH_FINALIZER_NAME]; has_annotation {
+				// wait for BMH to be deleted
+				if foundBMH, err := r.bmhExists(ctx, agent); err != nil || foundBMH {
+					if err != nil {
+						log.WithError(err).Warnf("failed to determine if BMH exists for agent")
+					}
+					log.Info("waiting for BMH to be deleted")
+					return &ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
+				}
+				// delete spoke node
+				spokeClient, err := r.spokeKubeClient(ctx, agent.Spec.ClusterDeploymentName)
+				if err != nil {
+					log.WithError(err).Error("failed to create spoke client, node will not be removed")
```

Comment:
Why don't we fail here (and below if DeleteNode fails)?

Reply:
It didn't seem like a recoverable error to me and if we fail we'd need to know when to give up. We'd have to implement some kind of tracker or timeout for retries.

It's also very possible that someone is deleting an SNO or an entire cluster and we won't be able to connect to delete the node.